service/frontend/handler.go

```diff
@@ -1511,15 +1513,38 @@ func (wh *WorkflowHandler) QueryWorkflow(ctx context.Context,
 	)
 
 	queryRequest.Execution.RunId = response.Execution.RunId
-	if len(response.StickyTaskList.GetName()) == 0 {
-		matchingRequest.TaskList = response.TaskList
-	} else if !clientFeature.SupportStickyQuery() {
-		// sticky enabled on the client side, but client chose to use non sticky, which is also the default
-		matchingRequest.TaskList = response.TaskList
-	} else {
+	if len(response.StickyTaskList.GetName()) != 0 && clientFeature.SupportStickyQuery() {
 		matchingRequest.TaskList = response.StickyTaskList
+		stickyDecisionTimeout := response.GetStickyTaskListScheduleToStartTimeout()
+		// using a clean new context in case customer provide a context which has
+		// a really short deadline, causing we clear the stickyness
+		stickyContext, cancel := context.WithTimeout(context.Background(), time.Duration(stickyDecisionTimeout)*time.Second)
+		matchingResp, err := wh.matching.QueryWorkflow(stickyContext, matchingRequest)
+		cancel()
+		if err == nil {
+			return matchingResp, nil
+		}
+		if yarpcError, ok := err.(*yarpcerrors.Status); !ok || yarpcError.Code() != yarpcerrors.CodeDeadlineExceeded {
```

Comment:
should it be !ok? Could you test this code path?

Reply:
this is tested by TestQueryWorkflow_StickyTimeout in integ test